Optimizations and QOL improvements

giorgi-o opened the pull request:
This PR contains the various touch-ups I have made to pdslib while working on the evaluation code. It mainly has performance improvements, and also contains a few misc touch-ups and fixes.

Each commit is self-contained, so it should be easy to review this PR commit-by-commit.

Major change:
- Replaced `HashSet<Uri>` with `UriSet<Uri>`, which is a wrapper around `Rc<HashSet<Uri>>`. This solves a major performance bottleneck, where `HashSet<Uri>`s were cloned quite often, and could contain thousands of URIs each. Plus, events with the same `trigger_uris` and `querier_uris` were duplicating this space, and many events often share the same `querier_uris`. Wrapping it in `Rc` is almost zero overhead, but makes cloning them a matter of incrementing a refcount.

Minor performance improvements:
- Allow EventStorage to implement its own relevant event selection. This is a small optimization to allow EventStorages to filter their own relevant events, instead of serializing 1000 events for a given epoch, only for 995 of them to be discarded as not relevant.
- Add foldhash hashing algorithm alternative. We already support ahash, and foldhash is the cutting-edge hashing algorithm with improvements across the board. I did not notice a significant difference in the evaluation code however.

Quality of life fixes:
- Make `HashMapEventStorage` and `HashMapFilterStorage` fields public, as these are mainly demo/stub classes so peeking inside helps with debugging and evaluation.
- Filters now implement Deserialize in addition to Serialize
- Implement BucketKey for all types, not just `u64`
- ...and a few other misc touch-ups.

giorgi-o committed:
Make HashMapFilterStorage fields public
Since this type is mainly used for debugging and evaluation, it makes sense to allow peeking inside

giorgi-o committed:
Make Filter trait require Debug

giorgi-o committed:
Add convenience getters for FilterId's epoch and uri

giorgi-o committed:
Implement BucketKey for all types, not just u64

giorgi-o committed:
Make a few useful struct fields public
This is useful for evaluation.
In particular, HashMapEventStorage is mainly used as a placeholder/for debugging, so making it inspectable is quite useful.

giorgi-o committed:
Allow EventStorage to implement its own relevant event selection
Adds a new EventStorage::relevant_events_for_epoch(), with a default implementation that matches the previous behaviour.
This means that the code is still backwards-compatible, and behaves exactly the same.
However, the EventStorage can now override the default implementation if it has a more optimised implementation.

The main use-case for this is HashMapEventStorage. Before, when requesting the relevant events for epoch 6, it would clone ALL the events for that epoch, and discard the irrelevant ones.
Now, it can filter them beforehand and only clone the relevant ones.

There is an argument to be made that EventStorage is a thin, simple trait and wouldn't benefit from extra functionality. However in my opinion, its sibling trait FilterStorage benefits a lot from being slightly more than an FFI wrapper, so there is precedent.

giorgi-o committed:
Filters implement Deserialize
This is useful if the filterstore backing storage is a database, that serializes on storage and deserializes on retrieval

giorgi-o committed:
Misc optimizations, changes and touch-ups

giorgi-o committed:
Add foldhash hashing algorithm alternative
Foldhash is the cutting-edge hashing algortithm that is leading in most performance benchmarks.
As we already have the scaffolding to swap out hashing algorithm based on a cargo feature with aHash, implementing this was a matter of copying aHash's compatibility-layer (dual-license Apache and MIT) and adapting it to foldhash.

Overall, I did not notice a significant speedup compared to ahash unfortunately.

giorgi-o committed:
Implement UriSet for storing URIs
In the evaluation, the main source of memory usage (often 100GB+) was the fact we store every Vec<Uri> as a seperate vector, essentially wasting memory as each event was storing its own two copies of its querier_uris/trigger_uris (which are often the same).
It is also a huge performance hit, as the overhead of cloning that vector repeatedly is huge.

This commit introduces a new data structure, UriSet, which is a thin rapper over Rc<HashSet<Uri>>.
This is based on 3 observations:
1. Event URIs are tested for the presence of an URI much more often than they are iterated over, so using a HashSet makes sense.
2. Often an event's trigger_uris and querier_uris are the same, so allowing them to share memory reduces PpaEvent's footprint by almost half.
3. An event's URIs are not modified after the event is created. Rc is a low-overhead, shared-ownership smart container which allows URIs.clone() to be an O(1) operation, with negligeable downsides.

Overall, the results are significant: when running the evaluation, memory usage is 1/6th of what it was before, and performance is 4x as fast for each workload, and 20x faster if we take into account that all the workloads can run in parallel due to memory no longer being a constraint.

tholop commented on src/events/traits.rs:
```diff
@@ -58,4 +58,20 @@ pub trait EventStorage {
+        epoch_id: &<Self::Event as Event>::EpochId,
+        relevant_event_selector: &impl RelevantEventSelector<Event = Self::Event>,
+    ) -> Result<impl Iterator<Item = Self::Event>, Self::Error> {
+        let iter = self
+            .events_for_epoch(epoch_id)?
+            .filter(|event| relevant_event_selector.is_relevant_event(event));
+        Ok(iter)
+    }
```
I'd add a comment inside, saying that this is the default implementation, which can be inefficient since it retrieves (and potentially clones) all events first.

giorgi-o replied:
Done in https://github.com/columbia/pdslib/pull/97/commits/4de3757828a35efccabb83bb82c0f4b45035c9d9

tholop commented on src/events/uri_set.rs:
```diff
@@ -0,0 +1,48 @@
+    fn into_iter(self) -> Self::IntoIter {
+        self.uris.iter()
+    }
+}
```
Would you be able to document a bit more why these traits are needed for UriSet? Like who is hashing a UriSet, or what's the difference between the two IntoIterator?

giorgi-o replied:
Done in https://github.com/columbia/pdslib/pull/97/commits/cae4834bdbc33d7a9877313278da7be4b2f17c59. Note that one is `IntoIterator`, the other is `From` with an `IntoIterator` bound.

tholop commented on src/pds/cross_report.rs:
```diff
@@ -247,9 +246,12 @@ impl<U: Uri> AttributionObject<PpaHistogramRequest<U>> {
+
+                event_values.retain(|(event, _)| {
+                    !epoch_relevant_events_set.contains(event)
+                });
```
This is pretty convoluted, with two steps and some double negation about what gets dropped. Also probably slower than the previous code that only iterated over a small set of events (current epoch + relevant).

It seems to be coming from the new Vec type for `event_values` instead of the previous HashMap. I get that Vec might simplify things further down the line (e.g. in ppa_histogram), but inside the current function a HashMap looks pretty natural. 

Is there a way to reconcile both approaches, perhaps by converting the HashMap to a vector at the end? Or even by keeping event values as vectors, but grouped by epoch (that might not be possible with the current shape of `event_values`)? It's okay otherwise, just curious.

giorgi-o replied:
The main reason for making `event_values` a vec is because it is iterated on twice (in both calls to `map_events_to_buckets()`), but only deleted from in the OOB case. I figured iteration speed was more important to optimize for.

However I agree the logic was a bit clunky and hard to follow. I've cleaned it up in https://github.com/columbia/pdslib/pull/97/commits/c194f96af165bbf13a9aa0e26ad4876ec02f6c2a.

tholop reviewed (approved):
Thanks a lot @giorgi-o! The changes make a lot of sense, and the commits were a breeze to review. I'd just recommend two changes before you merge:
- Some minor documentation and potential simplifications, see my comments inline.
- I think we should drop the whole foldhash commit. You say it doesn't improve performance much in practice, but it adds a dependency and quite a lot of new code in src/utils, so it's just bloat. It was an interesting experiment, and we can keep the code somewhere like this PR or some other branch, but I don't think it's worth incorporating into pdslib.

giorgi-o committed:
improve EventStorage::relevant_events_for_epoch() documentation

giorgi-o committed:
Revert "Add foldhash hashing algorithm alternative"

This reverts commit 594f4934b66bc76483292800783e81f910dbbf11.
Foldhash did not add any meaningful speed improvement, but required a lot of code, adding bloat to pdslib.

giorgi-o committed:
Document UriSet methods and remove Hash impl

giorgi-o committed:
Update dependencies, lockfile, and edition

giorgi-o committed:
Optimize PpaHistogramEvent::event_values()
Take the max() in O(n) time instead of sorting.
Note this removes the console warning when an event has an invalid histogram index. That's fine because bucket_key() also warns in that case.

giorgi-o committed:
Clean up cross_report.rs
Make get_report() easier to read through and the logic clearer.
Also remove all OOB events in one go, instead of doing it every iteration.
